contracts/staking/ServiceStakingBase.sol

```diff
@@ -541,38 +593,78 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
+            // The index must always exist if the service is currently staked, otherwise it never existed or was evicted
+            uint256 idx;
+            for (; idx < serviceIds.length; ++idx) {
+                if (serviceIds[idx] == serviceId) {
+                    break;
+                }
+            }
 
-        // Check if the service is staked
-        if (sInfo.tsStart == 0) {
-            revert ServiceNotStaked(serviceId);
+            // The service was already evicted
+            if (idx == serviceIds.length) {
+                revert ServiceNotFound(serviceId);
+            }
+
+            // Evict the service from the set of staked services
+            setServiceIds[idx] = setServiceIds[setServiceIds.length - 1];
+            setServiceIds.pop();
+
+            emit ServiceEvicted(serviceId, sInfo.owner, sInfo.multisig, inactivity);
         }
+    }
 
+    /// @dev Calculates service staking reward during the last checkpoint period.
+    /// @param serviceId Service Id.
+    /// @return reward Service reward.
+    function calculateServiceStakingLastReward(uint256 serviceId) public view returns (uint256 reward) {
         // Calculate overall staking rewards
         (uint256 lastAvailableRewards, uint256 numServices, uint256 totalRewards, uint256[] memory eligibleServiceIds,
-            uint256[] memory eligibleServiceRewards, , ) = _calculateStakingRewards();
+            uint256[] memory eligibleServiceRewards, , , ) = _calculateStakingRewards();
 
         // If there are eligible services, proceed with staking calculation and update rewards for the service Id
         for (uint256 i = 0; i < numServices; ++i) {
             // Get the service index in the eligible service set and calculate its latest reward
             if (eligibleServiceIds[i] == serviceId) {
                 // If total allocated rewards are not enough, adjust the reward value
                 if (totalRewards > lastAvailableRewards) {
-                    reward += (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
+                    reward = (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
                 } else {
-                    reward += eligibleServiceRewards[i];
+                    reward = eligibleServiceRewards[i];
                 }
                 break;
             }
         }
     }
 
+    /// @dev Calculates overall service staking reward at current timestamp.
+    /// @param serviceId Service Id.
+    /// @return reward Service reward.
+    function calculateServiceStakingReward(uint256 serviceId) external view returns (uint256 reward) {
+        // Get current service reward
+        ServiceInfo memory sInfo = mapServiceInfo[serviceId];
+        reward = sInfo.reward;
+
+        // Check if the service is staked
+        if (sInfo.tsStart == 0) {
+            revert ServiceNotStaked(serviceId);
+        }
+
+        reward += calculateServiceStakingLastReward(serviceId);
+    }
```

Comment:
@mariapiamo it could be very intense to try to emit all the services and rewards during the Checkpoint() emit. We have added one more view reward function - one can always check a reward during a specific period specifying the block number. So there is a function to calculate the last reward, and there's a function to calculate the overall reward.

Reply:
I see that can be very intense and I know about that view method. But rn there is no way to use that method to create a query

Reply:
Does the separation of view functions into `calculateServiceStakingReward()` and `calculateServiceStakingLastReward()` work for that purpose?

Reply:
No, we need events. It is super intense trying to emit all the services and rewards during the Checkpoint() ?

Reply:
@mariapiamo why do you want so much info in the events? We can call a view function just fine?

Reply:
@DavidMinarsch rn there is no way to use that method to create sql queries